### PR TITLE
Loans modal [PAR-13]

### PR DESCRIPTION
**Pull request by lemniscate8**
Loans request modal and some loans management flow on the friend's library page.

### CODE REVIEW

**Comment by lemniscate8** [comment]
@Hashiphoto it seems that trying to use the AuthContext might be crashing the page. Am I using it wrong?
Screenshot 2020-05-08 13 46 37

**Comment by Hashiphoto** [comment]
That seems to be caused when the page changes before a promise has resolved, leading the code to update the state for a component that doesn't exist on the page anymore. I can implement a fix for it soon. https://www.debuggr.io/react-update-unmounted-component/

**Comment by Hashiphoto** [comment]
Do you think we should do that in this PR or wait on it?

File: src/ourtypes.ts [hunk 1]
```diff
@@ -5,10 +5,12 @@ export interface Friend {
   status: "requested" | "friends";
 }
 
-export interface Loan {
+export interface Loan extends LoanRequest {
   id: string;
   owner_id: string;
+  owner_contact: string;
   requester_id: string;
+  requester_contact: string;
```
Comment (Hashiphoto) [inline src/ourtypes.ts:13]: Ah, right I forgot about the contract info. I'll add that to the DB Schema

**Review by Hashiphoto — APPROVED** [review]
Looks good!

**Comment by lemniscate8** [comment]
I implemented that `isMounted` pattern with the ref, but it still seems like something is still silently crashing the page. I'm getting navigation to the `/` route then immediate redirection to `/library` when trying to go to `/library/#number`.

File: src/FriendLibraryPage.tsx [hunk 2]
```diff
@@ -32,83 +34,108 @@ const FriendLibraryPage: React.FC = () => {
...
       })
-      .finally(() => {
-        setLoansAreLoaded(true);
-      });
-  }, []);
+        .then((res) => res.json())
+        .then((res: Loan) => {
+          setBooks(
+            books.map((book) =>
+              book.id !== loan.book_id ? book : { ...book, loan: res }
```
Comment (lemniscate8) [inline src/FriendLibraryPage.tsx:73]: I think this pattern's a good way to update local data only if the server sends back a positive response.

**Pull request merged**